File: app/src/main/java/org/blitzortung/android/data/provider/blitzortung/MapBuilderFactory.java
package org.blitzortung.android.data.provider.blitzortung;

import org.blitzortung.android.data.beans.DefaultStrike;
import org.blitzortung.android.data.beans.Station;
import org.blitzortung.android.data.beans.StrikeAbstract;
import org.blitzortung.android.data.provider.blitzortung.generic.Consumer;
import org.blitzortung.android.data.provider.blitzortung.generic.LineSplitter;
import org.blitzortung.android.util.TimeFormat;

import java.util.Map;

public class MapBuilderFactory {

    private final LineSplitter strikeLineSplitter;
    private final LineSplitter stationLineSplitter;

    public MapBuilderFactory() {
        this(text -> text.split(" "), new StationLineSplitter());
    }

    public MapBuilderFactory(LineSplitter strikeLineSplitter, LineSplitter stationLineSplitter) {
        this.strikeLineSplitter = strikeLineSplitter;
        this.stationLineSplitter = stationLineSplitter;
    }

    public MapBuilder<StrikeAbstract> createAbstractStrikeMapBuilder() {
        return new MapBuilder<StrikeAbstract>(strikeLineSplitter) {

            DefaultStrike.DefaultStrikeBuilder strikeBuilder = DefaultStrike.builder();

            @Override
            protected void prepare(String[] fields) {
                strikeBuilder.timestamp(TimeFormat.parseTimestampWithMillisecondsFromFields(fields));
            }

            @Override
            protected void setBuilderMap(Map<String, Consumer> keyValueBuilderMap) {
                keyValueBuilderMap.put("pos", values -> {
                    strikeBuilder.longitude(Float.parseFloat(values[1]));
                    strikeBuilder.latitude(Float.parseFloat(values[0]));
                    strikeBuilder.altitude(Integer.parseInt(values[2]));
                });
                keyValueBuilderMap.put("str", values -> strikeBuilder.amplitude(Float.parseFloat(values[0])));
                keyValueBuilderMap.put("dev", values -> strikeBuilder.lateralError(Integer.parseInt(values[0])));
                keyValueBuilderMap.put("sta", values -> strikeBuilder.stationCount((short) values.length));
            }

            @Override
            public StrikeAbstract build() {
                return strikeBuilder.build();
            }
        };
    }

    public MapBuilder<Station> createStationMapBuilder() {
        return new MapBuilder<Station>(stationLineSplitter) {

            private final Station.StationBuilder stationBuilder = Station.builder();

            @Override
            protected void prepare(String[] fields) {
                stationBuilder.name("");
            }

            @Override
            protected void setBuilderMap(Map<String, Consumer> keyValueBuilderMap) {
                keyValueBuilderMap.put("city", values -> stationBuilder.name(values[0].replace("\"", "")));
                keyValueBuilderMap.put("pos", values -> {
                    stationBuilder.longitude(Float.parseFloat(values[1]));
                    stationBuilder.latitude(Float.parseFloat(values[0]));
                });
                keyValueBuilderMap.put("last_signal", values -> {
                    String dateString = values[0].replace("\"", "").replace("-", "").replace(" ", "T");
                    stationBuilder.offlineSince(TimeFormat.parseTime(dateString));
                });
            }

            @Override
            protected Station build() {
                return stationBuilder.build();
            }
        };
    }
}

File: app/src/main/java/org/blitzortung/android/data/provider/blitzortung/MapBuilderFactory.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.data.provider.blitzortung

import javax.inject.Inject
import javax.inject.Singleton
import org.blitzortung.android.data.beans.DefaultStrike
import org.blitzortung.android.data.beans.Strike
import org.blitzortung.android.util.TimeFormat

class MapBuilderFailedException(message: String) : Throwable(message)

@Singleton
class MapBuilderFactory(
    private val strikeLineSplitter: (String) -> Array<String>,
) {
    @Inject
    constructor() : this(::lineSplitter)

    fun createStrikeMapBuilder(): MapBuilder<Strike> {
        return object : MapBuilder<Strike>(strikeLineSplitter) {
            init {
                setBuilderMap()
            }

            var longitude: Double = 0.0
            var latitude: Double = 0.0
            var amplitude: Float = 0f
            private var timestamp: Long = 0
            private var lateralError: Int = 0
            private var altitude: Int = 0
            private var stationCount: Short = 0

            override fun prepare(fields: Array<String>) {
                if (fields.size > 2) {
                    timestamp = TimeFormat.parseTimestampWithMillisecondsFromFields(fields)
                } else {
                    throw MapBuilderFailedException("not enough values for time parsing")
                }
            }

            fun setBuilderMap() {
                keyValueBuilderMap["pos"] = { values ->
                    if (values.size >= 3) {
                        longitude = values[1].toDouble()
                        latitude = values[0].toDouble()
                        altitude = values[2].toInt()
                    } else {
                        throw MapBuilderFailedException("not enough values for location parsing")
                    }
                }
                keyValueBuilderMap["str"] = { values -> amplitude = java.lang.Float.parseFloat(values[0]) }
                keyValueBuilderMap["dev"] = { values -> lateralError = Integer.parseInt(values[0]) }
                keyValueBuilderMap["sta"] = { values -> stationCount = values.size.toShort() }
            }

            override fun build(): Strike {
                return DefaultStrike(
                    timestamp,
                    longitude,
                    latitude,
                    altitude,
                    amplitude,
                    stationCount,
                    lateralError.toDouble(),
                )
            }
        }
    }
}

fun lineSplitter(text: String): Array<String> {
    return text.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
}
